onnxruntime/python/tools/symbolic_shape_infer.py

```diff
@@ -2385,31 +2385,35 @@ def _infer_PythonOp(self, node):  # noqa: N802
         output_tensor_ranks = get_attribute(node, "output_tensor_ranks")
         assert output_tensor_ranks
 
-        # set the context output separately.
-        # The first output is autograd's context.
+        from onnxruntime.training.ortmodule._custom_autograd_function_exporter import PythonOpShapeInferStore
+
+        func_name = get_attribute(node, "func_name").decode()
+        shape_inferer = PythonOpShapeInferStore.get_shape_infer(func_name)
+
+        # Set the context output separately.
+        # The first output is torch.autograd.Function''s context.
         vi = self.known_vi_[node.output[0]]
         vi.CopyFrom(helper.make_tensor_value_info(node.output[0], onnx.TensorProto.INT64, []))
-        # TODO(pengwa): allow custom PythonOp shape inference.
-        if get_attribute(node, "func_name").decode() in [
-            "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation",
-            "onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep",
-        ]:
-            # PythonOp with func_name being "_InspectActivation" or "_IncrementStep" will behave exactly same as a normal
-            # PythonOp when execution. The only difference is that
-            # 1). those ops having same number of tensor inputs and tensor outputs;
-            # 2). and the i-th output tensor's shape is same as i-th input tensor's shape.
-            # Be noted, the count of custom autograd function might be bigger than output count, because there might
-            # be other non-tensor constant inputs (string, object, int, tuple, etc). But we did not make those constant
-            # inputs as ONNX op's input, instead they are stored in the attributes.
-            assert len(node.output) == len(node.input) + 1  # The output contains one extra context info.
-            for input_index in range(len(node.output) - 1):
-                # Process the i-th tensor outputs.
-                vi = self.known_vi_[node.output[input_index + 1]]
+
+        if shape_inferer is not None:
+            input_shapes = []
+            input_dtypes = []
+            for input_index in range(len(node.input)):
                 shape = self._get_shape(node, input_index)
-                output_dtype = self.known_vi_[node.input[input_index]].type.tensor_type.elem_type
-                vi.CopyFrom(helper.make_tensor_value_info(node.output[input_index + 1], output_dtype, shape))
+                input_shapes.append(shape)
+                input_dtype = self.known_vi_[node.input[input_index]].type.tensor_type.elem_type
+                input_dtypes.append(input_dtype)
+            output_shapes, output_dtypes = shape_inferer(node, input_shapes, input_dtypes)
```

Comment:
My understanding of the code is that it will just return a tuple (input_shapes, input_dtypes), so why are we calling shape_inferer? Is it because we can have different definition of infer_shape for different class?

Reply:
Yes, that's it. The two cases are simpler cases where input and output shared same size or data types, I changed the comment in case it is a bit misleading as we talked offline. Thank you!